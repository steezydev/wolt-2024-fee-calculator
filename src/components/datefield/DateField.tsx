import {
  autoCompleteDate,
  dateTimeStringToDate,
  isValidDate,
} from '@/helpers/date';
import { DateFieldProps } from '@/types/props/DateFieldProps';
import { format } from 'date-fns';
import { useEffect, useState } from 'react';

const DateField = ({ date, onDateChange, renderInput }: DateFieldProps) => {
  const [inputValue, setInputValue] = useState(
    format(date, 'dd.MM.yyyy HH:mm')
  );

  useEffect(() => {
    setInputValue(format(date, 'dd.MM.yyyy HH:mm'));
  }, [date]);

  // Set or revert date value on blur
  const handleBlur = () => {
    const autoCompletedValue = autoCompleteDate(inputValue);

    if (isValidDate(autoCompletedValue)) {
      onDateChange && onDateChange(dateTimeStringToDate(autoCompletedValue));
    } else {
      setInputValue(format(date, 'dd.MM.yyyy HH:mm'));
    }
  };

  return renderInput({
    value: inputValue,
    onChange: (e) => setInputValue(e.target.value),
    onBlur: () => handleBlur(),
  });
};

export default DateField;
