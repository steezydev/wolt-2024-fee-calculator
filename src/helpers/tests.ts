const localStorageMock = (function () {
    let store: Record<string, string> = {};
    return {
        getItem(key: string) {
            return store[key] || null;
        },
        setItem(key: string, value: string) {
            store[key] = value.toString();
        },
        clear() {
            store = {};
        },
    };
})();

const matchMediaMock = jest.fn().mockImplementation((query) => ({
    matches: query === '(prefers-color-scheme: dark)',
    media: query,
    onchange: null,
    addListener: jest.fn(),
    removeListener: jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
}));

export const mockContext = () => {
    beforeAll(() => {
        Object.defineProperty(window, 'localStorage', {
            value: {
                getItem: jest.fn().mockImplementation((key: string) => {
                    return localStorageMock.getItem(key);
                }),
                setItem: jest
                    .fn()
                    .mockImplementation((key: string, value: string) => {
                        localStorageMock.setItem(key, value);
                    }),
                clear: localStorageMock.clear,
            },
            writable: true,
        });

        window.matchMedia = matchMediaMock;
    });

    beforeEach(() => {
        (window.localStorage.getItem as jest.Mock).mockClear();
        (window.localStorage.setItem as jest.Mock).mockClear();
        matchMediaMock.mockClear();
    });
};

export const setupFakeDate = () => {
    beforeEach(() => {
        jest.useFakeTimers();
        jest.setSystemTime(new Date('2024-01-01'));
    });

    afterEach(() => {
        jest.useRealTimers();
    });
};

export const mockScrollTo = () => {
    beforeEach(() => {
        Element.prototype.scrollTo = jest.fn();
    });

    afterEach(() => {
        (Element.prototype.scrollTo as jest.Mock).mockRestore();
    });
};
