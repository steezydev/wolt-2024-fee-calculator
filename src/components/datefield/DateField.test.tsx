import DateField from './DateField';
import { dateFieldRenderTests } from './DateField.tests';

dateFieldRenderTests('DateField', DateField);
