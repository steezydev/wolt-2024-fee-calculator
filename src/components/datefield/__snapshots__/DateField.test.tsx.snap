// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateField when asked to render renders 1`] = `
<input
  onBlur={[Function]}
  onChange={[Function]}
  value="17.01.2024 12:00"
/>
`;

exports[`DateField when asked to render with additional class renders with additional class 1`] = `
<input
  onBlur={[Function]}
  onChange={[Function]}
  value="17.01.2024 12:00"
/>
`;
